### PR TITLE
[H2I-1] Setup MqttMessageListener and store data to InfluxDB

### PR DESCRIPTION
**Pull request by emiliyank**

### CODE REVIEW

File: src/main/java/de/wejago/hichi2influxDB/Hichi2influxDbApplication.java [hunk 1]
```diff
@@ -1,58 +1,13 @@
 package de.wejago.hichi2influxDB;
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/Hichi2influxDbApplication.java]: Only small caps in package! Just remove the DB

File: src/main/java/de/wejago/hichi2influxDB/controller/MqttController.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/v1/mqtt")
+public class MqttController {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/controller/MqttController.java]: We don't need a rest controller!

File: src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class InfluxDbService {
+    @Autowired
+    private InfluxDBConfig influxDBConfig;
+
+    public Point writePoint(SensorEntry sensorEntry) {
+        InfluxDBClient influxDBClient = influxDBConfig.dbConnection();
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java]: I don't think opening a connection on every new measurement and also API is not very efficient.
Also don't forget to close the connection afterwards.

File: src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+@Service
+public class MqttSubscriberService {
+    private final String PROTOCOL_TCP = "tcp";
+    private final String CLIENT_ID="spring-boot-mqtt-subscriber";
+    private static final Logger logger = LoggerFactory.getLogger(MqttSubscriberService.class);
+    @Value("${mqtt.broker.ip}")
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java]: Move those to a configuration class, too.

File: src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Value("${mqtt.password}")
+    private String password;
+
+    @Autowired
+    private InfluxDbService influxDbService;
+    private IMqttClient mqttClient;
+
+    public void connect() {
+        System.setProperty("org.slf4j.simpleLogger.defaultLogLevel", "debug");
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java]: Put that into a configuration file, too.

File: src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        try {
+            MemoryPersistence persistence = new MemoryPersistence();
+            String broker = this.generateBroker();
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java]: You don't need the "this." here.
Only use this when it is really needed!

File: src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+@Service
+public class MqttSubscriberService {
+    private final String PROTOCOL_TCP = "tcp";
+    private final String CLIENT_ID="spring-boot-mqtt-subscriber";
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java]: Name it like the application.

File: src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(MqttSubscriberService.class);
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java]: Use @Slf4j annotation on the class!

File: src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            logger.info("Successfully connected to MQTT client!");
+        } catch (MqttException e) {
+            logger.error("Error connecting to MQTT client!", e);
+        }
+    }
+
+    public void subscribe(String topic) {
+        try {
+
+            ObjectMapper objectMapper = new ObjectMapper();
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java]: Creation is expensive, let Spring inject it as a final variable!

File: src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    @Value("${mqtt.broker.port}")
+    private String brokerPort;
+
+    @Value("${mqtt.username}")
+    private String username;
+
+    @Value("${mqtt.password}")
+    private String password;
+
+    @Autowired
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java]: Avoid @Autowired
Make the variable private final and add a @RequiredArgsConstructor

File: src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        } catch (MqttException e) {
+            logger.error("Error connecting to MQTT client!", e);
+        }
+    }
+
+    public void subscribe(String topic) {
+        try {
+
+            ObjectMapper objectMapper = new ObjectMapper();
+            mqttClient.subscribe(topic, (t, message) -> {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/MqttSubscriberService.java]: I would put that lambda content in some extra class, like a SensorMessageSubscriber

File: src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import com.influxdb.client.domain.WritePrecision;
+import com.influxdb.client.write.Point;
+import de.wejago.hichi2influxDB.config.InfluxDBConfig;
+import de.wejago.hichi2influxDB.dto.SensorEntry;
+import java.time.Instant;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class InfluxDbService {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java]: I would stick with the Spring naming of a Repository here. What do you mean @cupakob

File: src/test/java/de/wejago/hichi2influxDB/unit/InfluxDbServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Mock
+    private InfluxDBConfig influxDBConfig;
+
+    @InjectMocks
+    private InfluxDbService influxDbService;
+
+    @BeforeEach
+    void setup() {
+        MockitoAnnotations.openMocks(this);
```
Comment (burberius) [inline src/test/java/de/wejago/hichi2influxDB/unit/InfluxDbServiceTest.java]: Use @ExtendWith(MockitoExtention.class) as class annotation. That's cleaner in the body in my opinion.

File: src/test/java/de/wejago/hichi2influxDB/unit/InfluxDbServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Test
+    void writePointShouldGenerateAndWriteMeasurementPoint() {
+        //GIVEN
+        SensorEntry sensorEntry = this.generateTestSensorEntry();
```
Comment (burberius) [inline src/test/java/de/wejago/hichi2influxDB/unit/InfluxDbServiceTest.java]: No this

File: src/test/java/de/wejago/hichi2influxDB/unit/InfluxDbServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        Point expectedPoint = generateExpectedMeasurementPoint(sensorEntry);
```
Comment (burberius) [inline src/test/java/de/wejago/hichi2influxDB/unit/InfluxDbServiceTest.java]: I would move that to the then part.

File: src/test/java/de/wejago/hichi2influxDB/unit/InfluxDbServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        when(influxDBConfig.dbConnection()).thenReturn(influxDBClient);
+        when(influxDBClient.getWriteApiBlocking()).thenReturn(writeApi);
+
+        //WHEN
+        influxDbService.writePoint(sensorEntry);
+
+        //THEN
+        verify(influxDBClient, times(1)).getWriteApiBlocking();
```
Comment (burberius) [inline src/test/java/de/wejago/hichi2influxDB/unit/InfluxDbServiceTest.java]: You don't verify the writing of the points!

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+        Whitebox.setInternalState(mqttSubscriberService, "brokerIp", "localhost");
+        Whitebox.setInternalState(mqttSubscriberService, "brokerPort", "1883");
+        Whitebox.setInternalState(mqttSubscriberService, "username", "user123");
+        Whitebox.setInternalState(mqttSubscriberService, "password", "test-pass");
+
+        // WHEN
+        mqttSubscriberService.connect();
+
+        // THEN
+        assertThat(output.getOut()).contains("Successfully connected to MQTT client!");
```
Comment (burberius) [inline src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: Relying on the output isn't a good solution. Verify it called the mock.

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+        Whitebox.setInternalState(mqttSubscriberService, "brokerIp", "localhost");
+        Whitebox.setInternalState(mqttSubscriberService, "brokerPort", "8080");
+        Whitebox.setInternalState(mqttSubscriberService, "username", "user123");
+        Whitebox.setInternalState(mqttSubscriberService, "password", "test-pass");
+
+        // WHEN
+        assertDoesNotThrow(() -> mqttSubscriberService.connect(), "The connect() method should not throw any exceptions");
+
+        // THEN
+        assertThat(output.getOut()).contains("Error connecting to MQTT client!");
```
Comment (burberius) [inline src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: We need some error handling and retry for the MQTT connection and probably also Influx.
I would say leave it for now and we will solve that in a later task, when it is running in docker.
Both problems should make the health state change after some retries for example.

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    @Test
+    public void testSubscribeWhenReceivingSkippedMessage() throws Exception {
+        //GIVEN
+        String testTopic = "test/topic";
+        testSensorEntry = this.generateTestSensorEntry();
+        mqttMessage = this.generateSkippedMqttMessage();
+        when(influxDbService.writePoint(testSensorEntry)).thenReturn(point);
```
Comment (burberius) [inline src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: Till line 101 it's the same code as in the next test, extract a method.

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+            IMqttMessageListener listener = invocation.getArgument(1);
+            listener.messageArrived(testTopic, mqttMessage);
+            return null;
+        }).when(mqttClient).subscribe(eq(testTopic), any(IMqttMessageListener.class));
+
+        //WHEN
+        mqttSubscriberService.subscribe(testTopic);
+
+        //THEN
+        verify(influxDbService, times(1)).writePoint(any());
```
Comment (burberius) [inline src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: You should somehow verify that the deserialization of json and conversion to a point works as expected.
You can also test only the small units of the conversion method. I expect JsonMapper to work as expected. ;-)

**Review by burberius — CHANGES_REQUESTED** [review]

File: src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java (hunk 12 above)
Comment (cupakob) [reply src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java]: yes and move it to another package, this one is the wrong package

File: src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java (hunk 3 above)
Comment (emiliyank) [reply src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java]: I don't think opening a connection on every new measurement and also API is not very efficient. - I didn't get this one.

Comment (burberius) [reply src/main/java/de/wejago/hichi2influxDB/service/InfluxDbService.java]: Every time you write a point/measurement you first open a new connection to InfluxDB and then open a new blocking API on it.
60 measurements per hour means 60 times opening those connections (and probably closing).

I would open the connection once in a variable and on every write check if it is still open, if not, reopen.

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java (hunk 17 above)
Comment (emiliyank) [reply src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: outdated

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java (hunk 18 above)
Comment (emiliyank) [reply src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: will leave it for next task

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java (hunk 19 above)
Comment (emiliyank) [reply src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: outdated

File: src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Configuration;
+
+@Slf4j
+@Configuration
+public class MqttClientConfig {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java:17]: Have a look at https://www.baeldung.com/configuration-properties-in-spring-boot

File: src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+    @Value("${mqtt.username}")
+    private String username;
+
+    @Value("${mqtt.password}")
+    private String password;
+
+    public IMqttClient setMqttClient() {
+        try {
+            MemoryPersistence persistence = new MemoryPersistence();
+            String broker = this.generateBroker();
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java]: this :-D
And inline the variable

File: src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+
+    public MqttConnectOptions setMqttConnectOptions() {
+        MqttConnectOptions connectOptions = new MqttConnectOptions();
+        connectOptions.setUserName(username);
+        connectOptions.setPassword(password.toCharArray());
+        connectOptions.setCleanSession(true);
+        return connectOptions;
+    }
+
+    private String generateBroker() {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java]: Better buildBrokerUrl

File: src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+
+        Point point = generateMeasurementPoint(sensorEntry);
+        writeApi.writePoint(point);
+
+        return point;
+    }
+
+    private void setConnectionIfNeeded() {
+        //set the connection only if it is empty of not active
+        if(influxDBClient == null || influxDBClient.ping() == false) {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java]: I would change it to !influxDbClient.ping()

File: src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+        setConnectionIfNeeded();
+        WriteApiBlocking writeApi = influxDBClient.getWriteApiBlocking();
+
+        Point point = generateMeasurementPoint(sensorEntry);
+        writeApi.writePoint(point);
+
+        return point;
+    }
+
+    private void setConnectionIfNeeded() {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java]: Better updateConnectionIfNeeded?

File: src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+@Repository
+@RequiredArgsConstructor
+public class InfluxDbRepository {
+    private final InfluxDBConfig influxDBConfig;
+
+    private InfluxDBClient influxDBClient;
+
+    public Point writePoint(SensorEntry sensorEntry) {
+        setConnectionIfNeeded();
+        WriteApiBlocking writeApi = influxDBClient.getWriteApiBlocking();
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java:22]: Does that need to be closed afterwards?

File: src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class MqttSubscriberService {
+    @Value("${mqtt.topic}")
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java]: That should be a configuration property, too

File: src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+    private String topic;
+
+    private final InfluxDbRepository influxDbRepository;
+
+    private final SensorMessageSubscriber sensorMessageSubscriber;
+
+    private final MqttClientConfig mqttClientConfig;
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java]: Isn't that injected by Spring?

File: src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+    @Value("${mqtt.broker.port}")
+    private String brokerPort;
+
+    @Value("${mqtt.username}")
+    private String username;
+
+    @Value("${mqtt.password}")
+    private String password;
+
+    public IMqttClient setMqttClient() {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java]: Should be getMqttClient
And with a @Bean annotation, you can even pass it into the Service as a bean, that way it can be mocked easily.

File: src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class MqttSubscriberService {
+    @Value("${mqtt.topic}")
+    private String topic;
+
+    private final InfluxDbRepository influxDbRepository;
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java]: unused

File: src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+            mqttClient = mqttClientConfig.setMqttClient();
+            connectOptions = mqttClientConfig.setMqttConnectOptions();
+            mqttClient.connect(connectOptions);
+            log.info("Successfully connected to MQTT client!");
+        } catch (MqttException e) {
+            log.error("Error connecting to MQTT client!", e);
+        }
+    }
+
+    public void subscribe() {
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java]: I think you can merge connect and subscribe, they belong together.
And then it can be the @PostConstruct method.
So the class gets leaner.

File: src/main/java/de/wejago/hichi2influx/service/SensorMessageSubscriber.java [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+@Service
+@RequiredArgsConstructor
+    public class SensorMessageSubscriber implements IMqttMessageListener {
+        private final ObjectMapper objectMapper;
+
+        private final InfluxDbRepository influxDbRepository;
+
+        @Override public void messageArrived(String s, MqttMessage mqttMessage) throws Exception {
+            try {
+                log.info("Received message: " + new String(mqttMessage.getPayload()));
```
Comment (burberius) [inline src/main/java/de/wejago/hichi2influx/service/SensorMessageSubscriber.java]: Extract mqttMessage.getPayload() as var and use it at both locations, currently you create 2 strings!

**Review by burberius — CHANGES_REQUESTED** [review]

File: src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java (hunk 21 above)
Comment (emiliyank) [reply src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java:17]: So the idea is to create a POJO only with the configurations and use it in the classes `InfluxDBConfig` and `MqttClientConfig`?

Comment (burberius) [reply src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java:17]: Right, and Spring can directly map the properties in the yaml/properties file to the fields.

File: src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java (hunk 26 above)
Comment (emiliyank) [reply src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java:22]: If we close it then we won't be able to use it again for the next write. So we can remove the `updateConnectionIfNeeded()` because it will be always needed.

Comment (burberius) [reply src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java:22]: I mean close/release the writeApi

File: src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java (hunk 28 above)
Comment (emiliyank) [reply src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java]: Yes and actually it was not used here anymore so I removed it.

File: src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java (hunk 26 above)
Comment (emiliyank) [reply src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java:22]: I could not find close() method for the `writeApi`. In the documentation and the sample project they are not closing it either.

Comment (burberius) [reply src/main/java/de/wejago/hichi2influx/repository/InfluxDbRepository.java:22]: OK

**Review by burberius — APPROVED** [review]

File: src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java (hunk 29 above)
Comment (emiliyank) [reply src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java]: I don't get the idea for a **@bean**. Currently the `MqttClientConfig` is a **@Configuration** which makes the whole class a bean. And we can mock it easily.

**Comment by cupakob** [comment]
@emiliyank please squash the commits (and maybe reword the final commit message)

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java (hunk 20 above)
Comment (emiliyank) [reply src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: We verify that we call the **writePoint()** method with a proper parameter. I guess this is enough, isn't it?

File: src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java [hunk 33]
```diff
@@ -0,0 +1,43 @@
...
+    public MqttConnectOptions getMqttConnectOptions() {
+        MqttConnectOptions connectOptions = new MqttConnectOptions();
+        connectOptions.setUserName(mqttProperties.getUsername());
+        connectOptions.setPassword(mqttProperties.getPassword().toCharArray());
+        connectOptions.setCleanSession(true);
+        return connectOptions;
+    }
+
+    private String buildBrokerUrl() {
+        return PROTOCOL_TCP + "://" + mqttProperties.getBrokerIp() + ":" + mqttProperties.getBrokerPort();
```
Comment (cupakob) [inline src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java]: Maybe better to use this:
```
UriComponents uriComponents = UriComponentsBuilder.newInstance()
      .scheme("http").host("www.baeldung.com").path("/junit-5").build();
```

File: src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java [hunk 34]
```diff
@@ -0,0 +1,41 @@
...
+@RequiredArgsConstructor
+public class MqttSubscriberService {
+
+    private final SensorMessageSubscriber sensorMessageSubscriber;
+
+    private final MqttClientConfig mqttClientConfig;
+
+    private final MqttProperties mqttProperties;
+
+    private IMqttClient mqttClient;
```
Comment (cupakob) [inline src/main/java/de/wejago/hichi2influx/service/MqttSubscriberService.java]: `mqttClient` and `connectOptions` can be converted to a local variables

File: src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java (hunk 20 above)
Comment (burberius) [reply src/test/java/de/wejago/hichi2influxDB/unit/MqttSubscriberServiceTest.java]: It's OK

File: src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java (hunk 29 above)
Comment (burberius) [reply src/main/java/de/wejago/hichi2influx/config/MqttClientConfig.java]: I'll show you later, how I mean it.
You can merge for now.

**Review by cupakob — APPROVED** [review]
